contracts/proposers/BurnAuction.sol

```diff
@@ -22,39 +32,64 @@ contract BurnAuction is Chooser {
     // auction is a relation of the best bid for each slot
     mapping(uint32 => Bid) public auction;
 
+    // deposits is current balances of coordinators that can used for next bids
+    mapping(address => uint256) public deposits;
+
     /**
      * @dev Event called when an coordinator beat the bestBid of the ongoing auction
      */
-    event NewBestBid(uint32 slot, address coordinator, uint128 amount);
+    event NewBestBid(uint32 slot, address coordinator, uint256 amount);
 
     /**
      * @dev RollupBurnAuction constructor
      * Set first block where the slot will begin
      * Initializes auction for first slot
      */
-    constructor() public {
+    constructor(address payable _donationAddress, uint256 _donationNumerator)
+        public
+    {
+        require(
+            donationNumerator <= DONATION_DENOMINATOR,
+            "BurnAuction, constructor: bad denominator"
+        );
+
         genesisBlock = getBlockNumber() + DELTA_BLOCKS_INITIAL_SLOT;
+        donationAddress = _donationAddress;
+        donationNumerator = _donationNumerator;
     }
 
     /**
      * @dev Receive a bid from an coordinator. If the bid is higher than the current bid it replace the existing bid
      */
-    function bid() external payable {
+    function bid(uint256 bidAmount) external payable {
         uint32 auctionSlot = currentSlot() + 2;
+        // if not initialized it must be 0
+        uint256 currentBidAmount = auction[auctionSlot].amount;
         require(
-            msg.value > auction[auctionSlot].amount,
-            "Your bid doesn't beat the current best"
+            bidAmount > currentBidAmount,
+            "BurnAuction, bid: less then current"
+        );
+        address coordinator = msg.sender;
+        require(
+            deposits[coordinator] + msg.value >= bidAmount,
+            "BurnAuction, bid: insufficient funds for bidding"
+        );
+        // update balances
+        // refund previous coordinator
+        updateBalance(auction[auctionSlot].coordinator, currentBidAmount, 0);
+        // update donation balance
+        updateBalance(
+            donationAddress,
+            bidAmount.mul(donationNumerator).div(DONATION_DENOMINATOR),
+            currentBidAmount.mul(donationNumerator).div(DONATION_DENOMINATOR)
         );
-        // refund, check 0 case (it means no bids yet for the auction, so no refund)
-        if (auction[auctionSlot].initialized && auction[auctionSlot].amount > 0)
-            auction[auctionSlot].coordinator.transfer(
-                auction[auctionSlot].amount
-            );
+        // update coordinator with remaining value
+        updateBalance(coordinator, msg.value, bidAmount);
```

Comment:
So is it correct that the coordinator:
1. specifies the `bidAmount` to bid
2. and choose to
- pay the bid with `msg.value`
- pay the bid with remaining deposits
- or pay the bid with `msg.value` and deposits combined
- or purely bid to deposit values for future use.

Reply:
Somehow easier reasoning can go like this:
1. msg.value adds up to the deposit.
2. bidAmount is paid from deposit.
3. remaining value stays in the deposit.

Reply:
Oh, that sounds easier to understand.